evap/evaluation/models_logging.py

```diff
@@ -0,0 +1,290 @@
+from collections import defaultdict, namedtuple
+from datetime import date, datetime, time
+from enum import Enum
+import itertools
+import logging
+import threading
+from json import JSONEncoder
+
+from django.conf import settings
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.contrib.postgres.fields import JSONField
+from django.core.exceptions import FieldDoesNotExist
+from django.db import models
+from django.db.models.signals import m2m_changed
+from django.dispatch import receiver
+from django.forms.models import model_to_dict
+from django.template.defaultfilters import yesno
+from django.template.loader import render_to_string
+from django.utils.formats import localize
+from django.utils.translation import gettext_lazy as _
+
+logger = logging.getLogger(__name__)
+
+
+class FieldActionType(str, Enum):
+    M2M_ADD = "add"
+    M2M_REMOVE = "remove"
+    M2M_CLEAR = "clear"
+    INSTANCE_CREATE = "create"
+    VALUE_CHANGE = "change"
+    INSTANCE_DELETE = "delete"
+
+
+FieldAction = namedtuple("FieldAction", "label type items")
+
+
+class InstanceActionType(str, Enum):
+    CREATE = "create"
+    CHANGE = "change"
+    DELETE = "delete"
+
+
+class LogJSONEncoder(JSONEncoder):
+
+    def default(self, obj):
+        if isinstance(obj, (date, time, datetime)):
+            return localize(obj)
+        return super().default(obj)
+
+
+def _choice_to_display(field, choice):  # does not support nested choices
+    return next(filter(lambda t: t[0] == choice, field.choices), (choice, choice))[1]
+
+
+def _field_actions_for_field(model, field_name, actions):
+    field = model._meta.get_field(field_name)
+    try:
+        label = getattr(field, "verbose_name", field_name).capitalize()
+    except FieldDoesNotExist:
+        label = field_name.capitalize()
+
+    for field_action_type, items in actions.items():
+        if field.many_to_many or field.many_to_one or field.one_to_one:
+            # convert item values from primary keys to string-representation for relation-based fields
+            related_objects = field.related_model.objects.filter(pk__in=items)
+            missing = len(items) - related_objects.count()
+            items = [str(obj) for obj in related_objects] + ["�"] * missing
```

Comment:
how about <deleted object> instead of �? then you don't need to read the ~~manual~~code to know what that symbol means.

Reply:
what instead?

Reply:
nice :D i meant `<deleted object>`, but github swallowed that.

Reply:
Hmm would we want extra css for that?

Reply:
if you ask me, no, in my opinion the `<>` are visual indicator enough.